main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -272,67 +289,69 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
-      val us2 = visitExp(exp2, env0)
-      Used.empty ++ us1 ++ us2
+      val env1 = env0.addApply.stopBuilding
+      val env2 = env0.resetApplies.stopBuilding
+      val us1 = visitExp(exp1, env1)
+      val us2 = visitExp(exp2, env2)
+      us1 and us2
 
     case Expression.Unary(_, exp, _, _, _) =>
-      visitExp(exp, env0)
+      visitExp(exp, env0.resetApplies.stopBuilding)
 
     case Expression.Binary(_, exp1, exp2, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      us1 ++ us2
+      val us1 = visitExp(exp1, env0.resetApplies.stopBuilding)
+      val us2 = visitExp(exp2, env0.resetApplies.stopBuilding)
+      us1 and us2
 
     case Expression.Let(sym, exp1, exp2, _, _, _) =>
       // Extend the environment with the variable symbol.
       val env1 = env0 + sym
 
       // Visit the two expressions under the extended environment.
-      val innerUsed1 = visitExp(exp1, env1)
-      val innerUsed2 = visitExp(exp2, env1)
+      val innerUsed1 = visitExp(exp1, env1.resetApplies.stopBuilding)
+      val innerUsed2 = visitExp(exp2, env1.resetApplies.stopBuilding)
 
       // Check for shadowing.
       val shadowedVar = shadowing(sym, env0)
 
       // Check if the let-bound variable symbol is dead in exp2.
       if (deadVarSym(sym, innerUsed2))
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym + UnusedVarSym(sym)
+        (innerUsed1 and innerUsed2 ++ shadowedVar) - sym + UnusedVarSym(sym)
       else
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym
+        (innerUsed1 and innerUsed2 ++ shadowedVar) - sym
 
     case Expression.LetRec(sym, exp1, exp2, _, _, _) =>
       // Extend the environment with the variable symbol.
       val env1 = env0 + sym
 
       // Visit the two expressions under the extended environment.
-      val innerUsed1 = visitExp(exp1, env1)
-      val innerUsed2 = visitExp(exp2, env1)
+      val innerUsed1 = visitExp(exp1, env1.resetApplies.stopBuilding)
+      val innerUsed2 = visitExp(exp2, env1.resetApplies.stopBuilding)
 
       // Check for shadowing.
       val shadowedVar = shadowing(sym, env0)
 
       // Check if the let-bound variable symbol is dead in exp1 and exp2.
       if (deadVarSym(sym, innerUsed1 ++ innerUsed2))
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym + UnusedVarSym(sym)
+        (innerUsed1 and innerUsed2 ++ shadowedVar) - sym + UnusedVarSym(sym)
       else
-        (innerUsed1 ++ innerUsed2 ++ shadowedVar) - sym
+        (innerUsed1 and innerUsed2 ++ shadowedVar) - sym
 
     case Expression.IfThenElse(exp1, exp2, exp3, _, _, _) =>
-      val us1 = visitExp(exp1, env0)
-      val us2 = visitExp(exp2, env0)
-      val us3 = visitExp(exp3, env0)
-      us1 ++ us2 ++ us3
+      val us1 = visitExp(exp1, env0.resetApplies.stopBuilding)
+      val us2 = visitExp(exp2, env0.resetApplies.stopBuilding)
+      val us3 = visitExp(exp3, env0.resetApplies.stopBuilding)
+      us1 and (us2 or us3)
```

Comment:
So I now intuitively understand the meaning of `and` and `or`. I think their names makes sense too.